llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp

```diff
@@ -31,15 +31,55 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
   const LLT s32 = LLT::scalar(32);
   const LLT s64 = LLT::scalar(64);
 
+  const LLT nxv1s8 = LLT::scalable_vector(1, s8);
+  const LLT nxv2s8 = LLT::scalable_vector(2, s8);
+  const LLT nxv4s8 = LLT::scalable_vector(4, s8);
+  const LLT nxv8s8 = LLT::scalable_vector(8, s8);
+  const LLT nxv16s8 = LLT::scalable_vector(16, s8);
+  const LLT nxv32s8 = LLT::scalable_vector(32, s8);
+  const LLT nxv64s8 = LLT::scalable_vector(64, s8);
+
+  const LLT nxv1s16 = LLT::scalable_vector(1, s16);
+  const LLT nxv2s16 = LLT::scalable_vector(2, s16);
+  const LLT nxv4s16 = LLT::scalable_vector(4, s16);
+  const LLT nxv8s16 = LLT::scalable_vector(8, s16);
+  const LLT nxv16s16 = LLT::scalable_vector(16, s16);
+  const LLT nxv32s16 = LLT::scalable_vector(32, s16);
+
+  const LLT nxv1s32 = LLT::scalable_vector(1, s32);
+  const LLT nxv2s32 = LLT::scalable_vector(2, s32);
+  const LLT nxv4s32 = LLT::scalable_vector(4, s32);
+  const LLT nxv8s32 = LLT::scalable_vector(8, s32);
+  const LLT nxv16s32 = LLT::scalable_vector(16, s32);
+
+  const LLT nxv1s64 = LLT::scalable_vector(1, s64);
+  const LLT nxv2s64 = LLT::scalable_vector(2, s64);
+  const LLT nxv4s64 = LLT::scalable_vector(4, s64);
+  const LLT nxv8s64 = LLT::scalable_vector(8, s64);
+
   using namespace TargetOpcode;
 
+  const std::initializer_list<LLT> scalarLegalTypes = {s32, sXLen};
```

Comment:
I am not sure how helpful splitting these out with variable is. They are only used a single time by the `G_ADD, G_SUB, G_AND, G_OR, G_XOR` instructions, but scalarLegalTypes and vExtendedLegalTypes make it sound like these types should be used for all opcodes. If you want to keep them as a separate variable instead of using them in place in `legalFor`, then maybe you can change the name to make it clear that the list only applies to the specific opcodes it is used for.

Take a look at RISCVISelLowering file starting at the line that defines `BoolVecVTs`. You can reuse that as inspiration and maybe build up a `AllVecLLTs`?

Reply:
I would do something like this:

```
const LLT AllScalableVecLLTs = { ... };
  getActionDefinitionsBuilder({G_ADD, G_SUB, G_AND, G_OR, G_XOR})
      .legalFor(ST.hasVInstructions() ? AllScalableVecLLTs : {s32, sXLen})
      .widenScalarToNextPow2(0)
      .clampScalar(0, s32, sXLen);
```

Reply:
That code won't compile. `legalFor` takes an initializer_list, but you can't create named initializer_lists that are reusable.

Reply:
```
      .legalFor(ST.hasVInstructions() ? AllScalableVecLLTs : {}) 
```

from my suggestion could be replaced with

```
.legalIf(all(ST.hasVInstructions(), typeInSet(AllScalableVecLLTs)))
```

Reply:
typeInSet also takes an initializer_list, and it take a type index argument.

Reply:
Ok maybe C++14 changed initializer_list behavior so that the underlying array lifetime gets extended.

```
The underlying array is a temporary array, in which each element is copy-initialized (except that narrowing conversions are invalid) from the corresponding element of the original initializer list. The lifetime of the underlying array is the same as any other [temporary object](https://www.enseignement.polytechnique.fr/informatique/INF478/docs/Cpp/en/cpp/language/lifetime.html%23Temporary_object_lifetime), except that initializing an initializer_list object from the array extends the lifetime of the array exactly like [binding a reference to a temporary](https://www.enseignement.polytechnique.fr/informatique/INF478/docs/Cpp/en/cpp/language/reference_initialization.html%23Lifetime_of_a_temporary) (with the same exceptions, such as for initializing a non-static class member). The underlying array may be allocated in read-only memory.
```